Add Debug Line Information Support to LLILC

adiaaida opened the pull request:
This change adds support to LLILC to send source line information to
LLVM and then on to the CLR EE to be used in debugging. This change
adds:

* Support to the Reader to capture IL offsets and set them as debug
  locations for instructions. To do so, this change adds calls to the
  DIBuilder to create debug information for functions, and then to set
  the DebugLoc information for each instruction. The debugger expects
  offsets for the following: the start of a function, when the stack is
  empty, and when we have a call instruction. To get this all piped
  through LLVM properly, we had to insert extra nops for instances where
  the stack is empty, but we were looking at a store for a local (this
  gets swallowed by the following instruction). We also required extra
  nops around branches to get information about breaks piped through
  correctly. Finally, the debugger expects a location of where the
  return MSIL instruction was, but it expects the actual return logic to
  be considered part of the epilog. This is handled by the CoreCLR jit
  by inserting a nop with the IL offset of the return instruction, and
  everything after the nop has IL offset of -3 (epilog). Additionally,
  we had to change the logic for creating nops. The logic we previously
  had created a DoNothing intrinsic, which LLVM throws away. We needed
  these nops to make it all the way through to the AsmPrinter, so we
  changed nop creation to insert an inline asm nop, which LLVM preserves.
* Adds an object loaded layer to capture native and IL offsets and send
  them to the EE.

adiaaida commented:
@AndyAyersMS @pgavlin @russellhadley PTAL

pgavlin commented on include/Reader/readerir.h:
```diff
@@ -1610,6 +1619,11 @@ class GenIR : public ReaderBase {
                                                 ///< helper function.
                                                 ///< This constant is from the
                                                 ///< legacy jit.
+  struct DbgInfo {
```
Nit: s/Dbg/Debug/ here and in `LLILCDbgInfo`

dnfclas commented:
__@adiaaida__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +395,107 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+
+          if (NumDebugRanges > 0) {
+
+            // Allocate OffsetMapping array
```
Why not just allocate this lazily to avoid looping over the debug info twice?

adiaaida replied:
How? OffsetMapping has to be allocated before we can start writing to it (and this is how it was in Ryujit when I did it originally).

pgavlin replied:
Yeah, disregard that comment--I misread the code below.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +395,107 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
 }
+
+template <typename ObjSetT, typename LoadResult>
+void ObjectLoadListener::operator()(llvm::orc::ObjectLinkingLayerBase::ObjSetHandleT ObjHandles,
```
The `llvm::` qualification here is unnecessary,

adiaaida replied:
got it.

pgavlin commented on include/Reader/readerir.h:
```diff
@@ -558,6 +560,12 @@ class GenIR : public ReaderBase {
 
   EHRegion *rgnAllocateRegion() override;
   EHRegionList *rgnAllocateRegionList() override;
+  void setDebugLocation(uint32_t CurrOffset, uint32_t IsCall) override;
```
Please change `IsCall` to a `bool`.

AndyAyersMS replied:
Or maybe a enum with `IsCall` and `IsStackEmpty` the two possible values? Then the uses are self-documenting.

pgavlin replied:
Andy's suggestion is better: I hadn't really realized that `IsCall == 0` meant that the call originated from an empty operand stack.

adiaaida replied:
I can do IsStackEmpty, but we only set debug locations if StackEmpty is true or if it's a call instruction. So I thought the one bool would cover it.

AndyAyersMS commented on lib/Reader/reader.cpp:
```diff
@@ -6410,6 +6410,11 @@ void ReaderBase::readBytesForFlowGraphNodeHelper(
 
+    // Set debug locations
+    if (ReaderOperandStack->empty()) {
+      setDebugLocation(CurrentOffset, 0);
```
Agree with Pat that that second arg needs to convey information at call sites. If you make it a bool then change this to 
```
const bool IsCall = false;
setDebugLocation(CurentOffset, IsCall);
```
or similar.

adiaaida replied:
Done.

AndyAyersMS commented on lib/Reader/reader.cpp:
```diff
@@ -8015,6 +8027,8 @@ void ReaderBase::msilToIR(void) {
     ++it;
   }
 
+  popDebugScope();
```
It seems odd to me that `popDebugScope()` is here in the base class but `push` is only in the derived class. I would have expected them both to be at the same level...

adiaaida replied:
This is being deleted as Pat pointed out that we never have more than one scope, so there's no point in having the stack.

pgavlin commented on lib/Reader/readerir.cpp:
```diff
@@ -5381,11 +5519,15 @@ void GenIR::nop() {
+    // This will survive lowering, but we may want to do something else that
+    // is cleaner
+    llvm::FunctionType *FTy = llvm::FunctionType::get(
+        llvm::Type::getVoidTy(*(JitContext->LLVMContext)), /*Variadic=*/false);
```
The preferred style is to avoid the inline comment and use a const local, e.g.

```
const bool Variadic = false;
llvm::FunctionType *FTy = llvm::FunctionType::get(
        llvm::Type::getVoidTy(*(JitContext->LLVMContext)), Variadic);
```

adiaaida replied:
done.

pgavlin commented on lib/Reader/readerir.cpp:
```diff
@@ -428,6 +437,22 @@ void GenIR::readerPrePass(uint8_t *Buffer, uint32_t NumBytes) {
+  // Setup function for emiting debug locations
+  DIFile *Unit = DBuilder->createFile(LLILCDbgInfo.TheCU->getFilename(),
+                                      LLILCDbgInfo.TheCU->getDirectory());
+  bool isOptimized = (JitContext->Flags & CORJIT_FLG_DEBUG_CODE) == 0;
```
s/isOptimized/IsOptimized/

adiaaida replied:
done.

pgavlin commented on lib/Reader/readerir.cpp:
```diff
@@ -428,6 +437,22 @@ void GenIR::readerPrePass(uint8_t *Buffer, uint32_t NumBytes) {
+  DISubprogram *SP = DBuilder->createFunction(
+      FContext, Function->getName(), StringRef(), Unit, LineNo,
+      createFunctionType(Function, Unit), Function->hasInternalLinkage(),
+      true /* definition */, ScopeLine, DINode::FlagPrototyped, isOptimized,
```
Same issue with the inline comment here as below.

adiaaida replied:
done.

AndyAyersMS commented on lib/Reader/reader.cpp:
```diff
@@ -6994,6 +7001,7 @@ void ReaderBase::readBytesForFlowGraphNodeHelper(
     case ReaderBaseNS::CEE_LDC_I4_8:
     case ReaderBaseNS::CEE_LDC_I4_M1:
     LOAD_CONSTANT:
+      nop();
```
I'm confused why we need a `nop` here.

adiaaida replied:
This is to make sure that the store of the constant gets an ILOffset. What I was seeing is that when we have a LDC followed by a call, the call's ILOffset would be used instead of the load's. The EE expects both the call's offset (with the CALL_INSTRUCTION sourcetype) and the load's offset because the stack is empty. Generally speaking, LDC's are at locations where the stack is empty, and the debugger expects il offsets for all points where the stack is empty. I could add a comment to make that clear.

AndyAyersMS replied:
Ah, I think I see what you're getting at ...certain MSIL operations generate operands, not instructions. So if we have an MSIL operation that only creates an operand, and it has a reportable debug location, and the following MSIL operation creates an instruction and has some other reportable debug location, that first debug location doesn't get recorded anywhere and is lost.

So arguably we only need this `nop` if we're at a stack empty point, right? And presumably `ldc` is the only case where we generate an operand and not an instruction?

You might be able to make this more robust in the main IL processing loop -- if you are at a stack empty, down at the bottom of the loop after we've generated the IR, if the last instruction in the current block doesn't have the stack empty debug info, you could generate a `nop` there. But if `ldc` is truly the only case of this then adding a comment here (and possibly only adding a `nop`\ for stack empty) is ok by me.

adiaaida replied:
I saw this with breaks too, which is where that nop for branches came from. I'm not sure how I would go about doing the check of if the last instruction in the current block doesn't have the stack empty debug info.

AndyAyersMS replied:
For control flow I'd imagine that LLVM might optimize away "branch to next" even in debug codegen mode, so you'd lose the debug location if that happened.

If you're confident the only special case here (aside from branches) is `ldc` then just add a comment explaining why. If there are more special cases then I'd consider trying to do something more general.

pgavlin commented on lib/Reader/readerir.cpp:
```diff
@@ -945,6 +971,118 @@ void GenIR::createSafepointPoll() {
+  EltTys.push_back(ReturnTy);
+
+  for (auto ParamIterator = FunctionTy->param_begin();
+       ParamIterator != FunctionTy->param_end(); ParamIterator++) {
```
Nit: lift the call to `param_end()` out of the loop or just use `for (Type *Ty : FunctionTy->params())`

adiaaida replied:
done.

pgavlin commented on lib/Reader/readerir.cpp:
```diff
@@ -945,6 +971,118 @@ void GenIR::createSafepointPoll() {
+  // for each param, etc. convert the type to DIType and add it to the array.
+  FunctionType *FunctionTy = F->getFunctionType();
+
+  unsigned i = 0;
```
I think that this is unused aside from the increment in the loop below.

adiaaida replied:
Yep, this was an artifact of some of my debugging. Removing.

pgavlin commented on lib/Reader/readerir.cpp:
```diff
@@ -945,6 +971,118 @@ void GenIR::createSafepointPoll() {
+    uint64_t Size = JitContext->TM->getDataLayout()->getPointerTypeSize(Ty);
+    uint64_t Align = JitContext->TM->getDataLayout()->getPrefTypeAlignment(Ty);
+    llvm::DIType *DbgTy = DBuilder->createPointerType(
+        convertType(Ty->getContainedType(0)), Size, Align);
```
Please replace `getContainedType(0)` with `getPointerElementType()`.

adiaaida replied:
done.

AndyAyersMS commented on lib/Reader/reader.cpp:
```diff
@@ -6564,6 +6569,7 @@ void ReaderBase::readBytesForFlowGraphNodeHelper(
       goto GEN_BRANCH;
 
     GEN_BRANCH:
+      nop();
```
Would love to see a comment here as to why we need a `nop`, even if it will just be there in a debug compile. I think I know why...maybe...

pgavlin replied:
FWIW, it might be nice to change the name of this function to `debug_nop` or similar to indicate that nothing is emitted unless we're compiling for debug.

adiaaida replied:
This was another instance of needed a nop to make the debugger happy, in particular with breaks. This was caught by one of the automated debugger tests that had breaks in it. Basically this nop maps to the il offset of where the curly brace source line is.

pgavlin commented on lib/Reader/readerir.cpp:
```diff
@@ -945,6 +971,118 @@ void GenIR::createSafepointPoll() {
+    return DbgTy;
+  }
+  // TODO: add support for aggregate types
+  return nullptr;
```
Is it OK to just return `nullptr` from this function? Do we understand the impact of doing so? I don't think that we should see FCAs in our function signatures at the moment in any case, so it might be better to just turn this into an `assert(UNREACHED)` or `llvm_unreachable()`.

adiaaida replied:
This isn't unreachable. There are other types that I do not have support for that do end up reaching this point. It hasn't proved to be any sort of problem yet.

pgavlin replied:
Are those types only aggregates? It would be nice to know exactly what we're missing here.

pgavlin replied:
Any more info here?

AndyAyersMS commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +395,107 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+    const void *ObjData = DebugObj.getData().data();
+    DIContext *DwarfContext = new DWARFContextInMemory(DebugObj);
+
+    // Use symbol info to iterate functions in the object.
```
Any idea how this will need to change when we have funclets running around? If not, please leave a TODO and open an issue for debug with EH.

adiaaida replied:
I don't know. I'll add a TODO.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +395,107 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+        ICorDebugInfo::OffsetMapping *OM;
+        // unsigned LastDebugEntry = 0;
+
+        if (DwarfContext) {
```
Can this condition ever be false? It looks like `DwarfContext` is only ever set once.

adiaaida replied:
I don't know. I got this from the IntelJITEventListener, and this is how they have it. I can try removing it and seeing if anything breaks.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +395,107 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+
+    // Get the address of the object image for use as a unique identifier
+    const void *ObjData = DebugObj.getData().data();
+    DIContext *DwarfContext = new DWARFContextInMemory(DebugObj);
```
Can this be stack-allocated? `DWARFContextInMemory DwarfContext(DebugObj);`

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +395,107 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+template <typename ObjSetT, typename LoadResult>
+void ObjectLoadListener::operator()(llvm::orc::ObjectLinkingLayerBase::ObjSetHandleT ObjHandles,
+                const ObjSetT &Objs, const LoadResult &LoadedObjInfos) {
+  int I = 0;
```
I think that this `I` is also unused except for a later increment.

adiaaida replied:
Done.

adiaaida replied:
Actually, nope. It's used on line 406.

pgavlin replied:
Ah, yep. Good catch.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +395,107 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+        unsigned NumDebugRanges = 0;
+        unsigned Offset = 0;
+        ICorDebugInfo::OffsetMapping *OM;
+        // unsigned LastDebugEntry = 0;
```
Remove this line.

adiaaida replied:
Which line?

pgavlin replied:
The commented out declaration.

adiaaida replied:
Done.

AndyAyersMS commented on lib/Reader/readerir.cpp:
```diff
@@ -945,6 +971,118 @@ void GenIR::createSafepointPoll() {
+  LLVMBuilder->SetCurrentDebugLocation(Loc);
+}
+
+llvm::DISubroutineType *GenIR::createFunctionType(llvm::Function *F,
```
We should leave a note somewhere about these dwarf type mappings.

The CLR/EE will never look at them, so (I assume) we have them because if we run on linux and use native debugging then gdb or similar is informed about the debug for the jitted code we can get some information when debugging without using the "standard" xplat approach -- though it is debatable how useful this will be is given that there is no ability to map back to source.

I'm fine with keeping this code around but we should add a note about its purpose and why it is ok for it to be incomplete.

adiaaida replied:
We have it because I needed something for the functiontype, and I wasn't sure if/how this was used throughout LLVM. I don't think it is necessary for us. I'm also not sure it's necessary for GDB for JIT (looking at the GDB EventListener)

AndyAyersMS commented on lib/Reader/readerir.cpp:
```diff
@@ -5381,11 +5519,15 @@ void GenIR::nop() {
-    Value *DoNothing = Intrinsic::getDeclaration(JitContext->CurrentModule,
-                                                 Intrinsic::donothing);
+    // This will survive lowering, but we may want to do something else that
+    // is cleaner
```
I think we'll eventually want a high-level NOP that doesn't get removed and gets lowered to the right platform-specific encoding, so perhaps you can add a todo and an ISSUE to implement that in LLVM.

adiaaida replied:
Adding a TODO.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +395,107 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+        uint64_t Size = SI->getSize();
+
+        unsigned LastDebugOffset = -1;
+        unsigned CurrentDebugEntry = 0;
```
Nit: I'd move this declaration closer to its use.

adiaaida replied:
Done

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +395,107 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+        unsigned LastDebugOffset = -1;
+        unsigned CurrentDebugEntry = 0;
+        unsigned NumDebugRanges = 0;
+        unsigned Offset = 0;
```
I think that this local is unused as well (it ends up shadowed by the declaration of `int Offset` on line 471.

adiaaida replied:
Done.

JosephTremoulet commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +395,107 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
   // do nothing
 }
+
+template <typename ObjSetT, typename LoadResult>
```
As a rule, template function definitions should be in headers rather than cpp, no?

pgavlin replied:
Only if they're intended for use in more than one translation unit. In this case, the function is local to LLILCJit.cpp, as so does not need to be defined in the header file. It might be better, in fact, if `ObjectLoadListener` was simply a forward declaration in LLILCJit.h and defined completely in LLILCJit.cpp.

JosephTremoulet commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +395,107 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+void ObjectLoadListener::operator()(llvm::orc::ObjectLinkingLayerBase::ObjSetHandleT ObjHandles,
+                const ObjSetT &Objs, const LoadResult &LoadedObjInfos) {
+  int I = 0;
+  for (const std::unique_ptr<object::ObjectFile> &PObj : Objs) {
```
Since this is written as implementing a contract where the pointer kind (and collection type) of the `ObjectFile` pointer collection is (are) abstracted away in the type parameter, seems like you may as well use `const auto &PObj` instead here, to accommodate other pointer kinds.

JosephTremoulet commented on include/Jit/LLILCJit.h:
```diff
@@ -19,11 +19,32 @@
+#include <vector>
+
+using namespace llvm;
+using namespace llvm::object;
```
You don't seem to be making use of these new #includes or using statements later in this file; should they be here?

JosephTremoulet commented:
I seem to have suggested moving `ObjectLoadListener::operator()` from the .cpp to the .h in one comment, and moving the #includes it relies on from the .h to the .cpp in another comment.  It might make the most sense to move just the iteration part (that is templated over the pointer kind and set type) into the header file, but pull out the guts of the method into a non-template method that just takes an `ObjectFile &` and a `LoadedObjectInfo &`, whose definition would stay in the .cpp.

adiaaida commented:
I'm pretty sure I got most of these comments. I have probably missed something. Please take a look and let me know.

@pgavlin, @JosephTremoulet I went with Joe's suggestion to split the operator() to the .h file and create a function for doing the actual debug information communication which is in the .cpp.

pgavlin commented on include/Jit/LLILCJit.h:
```diff
@@ -174,6 +180,32 @@ class NullResolver : public llvm::RuntimeDyld::SymbolResolver {
   }
 };
 
+class ObjectLoadListener {
```
I still think that this type should be moved into LLILCJit.cpp, as it's not used outside that file. Its presence here is only to satisfy the typedef on line 221--this whole definition can be replaced with a forward declaration (i.e. `class ObjectLoadListener;`), which is all that is necessary for the header file.

JosephTremoulet replied:
FWIW, that would be fine with me -- so long as the template declaration and definition are in the same file, there's no worry about referencing an undefined instantiation.

adiaaida replied:
done.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +407,100 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+  DIContext *DwarfContext = new DWARFContextInMemory(DebugObj);
+
+  // Use symbol info to iterate functions in the object.
+  // TODO: This may have to change when we have funclets
```
Should there be an issue for this?

adiaaida replied:
Adding to my list. I haven't gotten around to adding issues.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +407,100 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+
+  // Get the address of the object image for use as a unique identifier
+  const void *ObjData = DebugObj.getData().data();
+  DIContext *DwarfContext = new DWARFContextInMemory(DebugObj);
```
I think that the heap allocation here can be removed. Instead of `DIContext *DwarfContext = new DWARFContextInMemory(DebugObj);`, `DWARFContextInMemory DwarfContext(DebugObj)` should do.

adiaaida replied:
done.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +407,100 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+    SymbolRef::Type SymType;
+    if (SI->getType(SymType))
+      continue;
+    if (SymType == SymbolRef::ST_Function) {
```
Nit: maybe change this block to an early-continue style for consistency with the other checks (i.e. instead of `if (SymType == SymbolRef::ST_Function) { ... }`, make this `if (SymType != SymbolRef::ST_Function) continue;`)

adiaaida replied:
done.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +407,100 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+      unsigned NumDebugRanges = 0;
+      ICorDebugInfo::OffsetMapping *OM;
+
+      if (DwarfContext) {
```
I don't think that this check can every fail (it certainly can't if you change `DwarfContext` from a pointer to a value as suggested).

adiaaida replied:
done.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +407,100 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+
+        // Count offset entries. Will skip an entry if the current IL offset
+        // matches
+        // the previous offset.
```
Formatting: please move `the previous offset` to the preceding line.

adiaaida replied:
Thanks. Fixed. ccformat fail.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +407,100 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+        unsigned CurrentDebugEntry = 0;
+
+        if (NumDebugRanges > 0) {
+
```
Nit: extra blank line.

adiaaida replied:
done.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +407,100 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+          OM = (ICorDebugInfo::OffsetMapping *)Context->JitInfo->allocateArray(
+              SizeOfArray);
+
+          Begin = Lines.begin();
```
Isn't this redundant? `Begin` should already be positioned at `Lines.begin()`.

adiaaida replied:
done.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +407,100 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+
+            // We store info about if the instruction is being recorded because
+            // it is a call in the column field
+            int IsCall = (It->second).Column;
```
I'd change this to `bool IsCall = (It->second).Column == 1;` to avoid the comparison in the ternary on line 491.

adiaaida replied:
done.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +407,100 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+              OM[CurrentDebugEntry].ilOffset = LineNumber;
+              OM[CurrentDebugEntry].source =
+                  IsCall == 1 ? ICorDebugInfo::CALL_INSTRUCTION
+                              : ICorDebugInfo::STACK_EMPTY;
```
Is this always correct? I see a few other source types listed in corinfo.h. Do we not generate any of those other types yet?

adiaaida replied:
Yes. Ryujit only generates Call_Instruction and Stack_Empty, so we only generate those as well.

pgavlin commented on lib/Reader/reader.cpp:
```diff
@@ -6399,6 +6399,7 @@ void ReaderBase::readBytesForFlowGraphNodeHelper(
 
   while (CurrentOffset < fgNodeGetEndMSILOffset(Fg)) {
     CORINFO_RESOLVED_TOKEN ResolvedToken;
+    bool IsCall = false;
```
I think that having this up so high is confusing. I'd just inline the declaration at the sites that use it (at that point, I think the resulting declarations could also be made const).

adiaaida replied:
done.

JosephTremoulet commented on include/Jit/LLILCJit.h:
```diff
@@ -19,11 +19,17 @@
+#include "llvm/Object/ObjectFile.h"
+
+using namespace llvm;
+using namespace llvm::object;
```
Nit: Per [the LLVM coding standards](http://llvm.org/docs/CodingStandards.html#do-not-use-using-namespace-std), I think it's best not to put using statements in headers:

> In header files, adding a `using namespace XXX` directive pollutes the namespace of any source file that #includes the header.

adiaaida replied:
done.

JosephTremoulet commented on include/Jit/LLILCJit.h:
```diff
@@ -174,6 +180,32 @@ class NullResolver : public llvm::RuntimeDyld::SymbolResolver {
+  template <typename ObjSetT, typename LoadResult>
+  void operator()(llvm::orc::ObjectLinkingLayerBase::ObjSetHandleT ObjHandles,
+                  const ObjSetT &Objs, const LoadResult &LoadedObjInfos) {
+    for (const std::unique_ptr<object::ObjectFile> &PObj : Objs) {
```
Nit: I'd still prefer `const auto &PObj` here, since we're going to the trouble of abstracting the set type and pointer kind to match the abstraction in the calling code.

adiaaida replied:
done.

JosephTremoulet commented:
LGTM aside from a couple nits.

adiaaida commented:
I think I got everything. Please let me know if you have anything else.

JosephTremoulet commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -22,30 +22,66 @@
+  void operator()(llvm::orc::ObjectLinkingLayerBase::ObjSetHandleT ObjHandles,
+                  const ObjSetT &Objs, const LoadResult &LoadedObjInfos) {
+    for (const auto &PObj : Objs) {
+      int I = 0;
```
This needs to be initialized outside the `for` loop.

adiaaida replied:
fixed.

JosephTremoulet commented:
LGTM (once you push that last fix).  Thanks for the updates.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +431,95 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+
+      // Iterate through the debug entries and save IL offset, native
+      // offset, and
+      // source reason
```
Comment formatting is off--please move `source reason` into the preceeding line.

pgavlin commented on lib/Jit/LLILCJit.cpp:
```diff
@@ -394,3 +431,95 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
+
+    // Reset offset
+    LastDebugOffset = -1;
+    unsigned CurrentDebugEntry = 0;
```
Minor nit: I'd sink this declaration down to ~line 499 or so (immediately preceding `for (DILineInfoTable::iterator It = Begin; It != End; ++It) {`).

pgavlin commented on lib/Reader/reader.cpp:
```diff
@@ -7342,6 +7355,10 @@ void ReaderBase::readBytesForFlowGraphNodeHelper(
+      // Insert a nop for debug and set return's IL offset to signify
+      // that it is part of the epilogue
+      nop();
+      setDebugLocation(ICorDebugInfo::EPILOG, false);
```
Nit: please life the constant `false` into a local similar to the other call sites: `const bool IsCall = false`.

pgavlin commented:
LGTM modulo two nits and a request that we add more info about which types we're missing in `GenIR::convertType`.

adiaaida commented:
Added a comment about which types we're missing and fixed all the comments above. FWIW, we are missing LabelTy, MetadataTy, X86_MMXTy, FunctionTy, StructTy, ArrayTy, VectoryTy

pgavlin commented:
Do you know which of those we actually see at runtime?

adiaaida commented:
Definitely StructTy. Not sure about the others. When I was running and printing things, I was only running on a small number of functions. I haven't run it on anything terribly complicated.

adiaaida commented:
@dotnet-bot test this please
